compiler/next/include/chpl/types/CompositeType.h

```diff
@@ -36,90 +36,96 @@ namespace types {
   types.
 
+  This class itself does not represent the field types. Instead, it
+  uses the uAST ID of the class/record/union. For a generic type,
+  it also uses a SubstitutionsMap.
+
+  (Storing field types in this class would present significant challenges
+   for computing this type in an immutable way within the query framework). 
  */
 class CompositeType : public Type {
  public:
-  struct FieldDetail {
-    UniqueString name;
-    bool hasDefaultValue = false;
-    const uast::Decl* decl = nullptr;
-    QualifiedType type;
-
-    FieldDetail(UniqueString name,
-                bool hasDefaultValue,
-                const uast::Decl* decl,
-                QualifiedType type)
-      : name(name), hasDefaultValue(hasDefaultValue), decl(decl), type(type) {
-    }
-    bool operator==(const FieldDetail& other) const {
-      return name == other.name &&
-             hasDefaultValue == other.hasDefaultValue &&
-             decl == other.decl &&
-             type == other.type;
-    }
-    bool operator!=(const FieldDetail& other) const {
-      return !(*this == other);
-    }
-    size_t hash() const {
-      return chpl::hash(name, hasDefaultValue, decl, type);
-    }
-
-    void stringify(std::ostream& ss, chpl::StringifyKind stringKind) const {
-      name.stringify(ss, stringKind);
-      //TODO: determine the proper way to do this
-      //decl.stringify(ss, stringKind);
-      type.stringify(ss, stringKind);
-    }
-
-    /// \cond DO_NOT_DOCUMENT
-    DECLARE_DUMP;
-    /// \endcond DO_NOT_DOCUMENT
-  };
- protected:
-  // TODO: add fields and accessors for a QualifiedType per field
+  /**
+    \rst
+
+    A SubstitutionsMap stores a mapping from uAST ID of a field or formal
+    argument to the QualifiedType for it.
+
+    When a generic type or function is instantiated, the SubstitutionsMap
+    describes the types that are used to instantiate it. For example,
+    if we have
+
+    .. code-block:: chapel
+
+        record R {
+          type t;
+        }
+
+    Then ``R(int)`` is the RecordType with SubstitutionsMap containing
+    a mapping from ``t`` to ``int``.
+
+    Similarly, if we have
+
+    .. code-block:: chapel
+
+        proc f(a, param b) { }
+
+    Then ``f(1.0, 2)`` creates an instantiation of ``f`` with the
+    SubstitutionsMap containing a mapping from ``a`` to ``real`` and a mapping
+    from ``b`` to ``int``.
+
+    \endrst
 
+   */
+  using SubstitutionsMap = std::unordered_map<ID, types::QualifiedType>;
+  using SubstitutionPair = std::pair<ID, QualifiedType>;
+  using SortedSubstitutions = std::vector<SubstitutionPair>;
+
+ protected:
   ID id_;
   UniqueString name_;
-  std::vector<FieldDetail> fields_;
 
   // Is this CompositeType representing an instantiation?
   // If so, what is the generic CompositeType that was instantiated?
+  // This is stored here to keep the canPass code more efficient.
   const CompositeType* instantiatedFrom_ = nullptr;
```

Comment:
Shouldn't this be an `ID` and not a pointer?

Reply:
It can be a pointer as long as we can fully resolve the "instantiated from" type before the pointer type. (I.e. recursion should not be possible with this).  I think having it be a pointer allows things to be more efficient (since we don't have to look up by ID etc). There is the same situation with the parent type for classes.